pinot-common/src/main/java/com/linkedin/pinot/common/lineage/SegmentMergeLineage.java

```diff
@@ -0,0 +1,313 @@
+    _parentGroupToChildrenGroupsMap = new HashMap<>();
+    _levelToGroupToSegmentsMap = new HashMap<>();
+  }
+
+  public SegmentMergeLineage(String tableNameWithType, Map<String, List<String>> segmentGroupLineageMap,
+      Map<Integer, Map<String, List<String>>> levelToGroupToSegmentMap, int version) {
+    _tableNameWithType = tableNameWithType;
+    _parentGroupToChildrenGroupsMap = segmentGroupLineageMap;
+    _levelToGroupToSegmentsMap = levelToGroupToSegmentMap;
+  }
+
+  public String getTableName() {
+    return _tableNameWithType;
+  }
+
+  public static SegmentMergeLineage fromZNRecord(ZNRecord record) {
+    String tableNameWithType = record.getId();
+    int version = record.getVersion();
+    Map<String, List<String>> segmentGroupLineageMap = record.getListFields();
+
+    Map<Integer, Map<String, List<String>>> groupToSegmentsMap = new HashMap<>();
+    for (Map.Entry<String, Map<String, String>> entry : record.getMapFields().entrySet()) {
+      String levelKey = entry.getKey();
+      Integer level = Integer.parseInt(levelKey.substring(LEVEL_KEY_PREFIX.length()));
+      Map<String, List<String>> groupToSegmentsForLevel = new HashMap<>();
+      for (Map.Entry<String, String> groupEntry : entry.getValue().entrySet()) {
+        String groupId = groupEntry.getKey();
+        String segmentsString = groupEntry.getValue();
+        List<String> segments = Arrays.asList(segmentsString.split(SEGMENT_DELIMITER));
+        groupToSegmentsForLevel.put(groupId, new ArrayList<>(segments));
+      }
+      groupToSegmentsMap.put(level, groupToSegmentsForLevel);
+    }
+    return new SegmentMergeLineage(tableNameWithType, segmentGroupLineageMap, groupToSegmentsMap, version);
+  }
+
+  public ZNRecord toZNRecord() {
+    ZNRecord record = new ZNRecord(_tableNameWithType);
+    record.setListFields(_parentGroupToChildrenGroupsMap);
+    Map<String, Map<String, String>> groupToSegmentsMap = new HashMap<>();
+
+    for (Map.Entry<Integer, Map<String, List<String>>> entry : _levelToGroupToSegmentsMap.entrySet()) {
+      String key = LEVEL_KEY_PREFIX + entry.getKey();
+      Map<String, String> groupSegmentsForLevel = new HashMap<>();
+      for (Map.Entry<String, List<String>> groupEntry : entry.getValue().entrySet()) {
+        String groupId = groupEntry.getKey();
+        String segments = String.join(SEGMENT_DELIMITER, groupEntry.getValue());
+        groupSegmentsForLevel.put(groupId, segments);
+      }
+      groupToSegmentsMap.put(key, groupSegmentsForLevel);
+    }
+    record.setMapFields(groupToSegmentsMap);
+
+    return record;
+  }
+
+  /**
+   * Add segment merge lineage information
+   *
+   * @param groupId a group id
+   * @param currentGroupSegments a list of segments that belongs to the group
+   * @param childrenGroups a list of children groups that the current group covers
+   */
+  public void addSegmentGroup(String groupId, List<String> currentGroupSegments, List<String> childrenGroups)
```

Comment:
`addSegmentGroup` will be called in 2 places.

1. Regular segment upload (via pbnj) : for original segments, we keep the group level 0 with `children = null`.
2. Minion merge task calling Controller API(will be added with the other pr): in this case, minion job needs to provide the children groups and call controller api to update segment merge lineage. 

I tried to keep API parameter as simple as possible because the minion task has limited information (although we can pass more information to task as task config). I think that we can revisit this issue for the pr for controller api.